<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html lang="en" xml:lang="en" xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="content-type" content="application/xhtml+xml; charset=UTF-8" />
    <title>Fast Alignment Search Tool suite</title>
    <meta name="keywords" content="sirFAST, sir FAST, Compbio, Computational biology, SFU, Short
      Read Aligner, next gen sequencing" />
    <link rel="stylesheet" type="text/css" href="css/style.css" />
  </head>
  <body>
    <div id="wrapper">
      <div id="header">
        <h1> sirFAST </h1>
        <h2> Fast and Accurate Mapping of Complete Genomics Reads </h2>
      </div>
      <div id="contentwrapper">
        <div id="main"><a href="http://sirfast.sourceforge.net/">Home</a> &gt;
          Manual<br />
          <br />
          <hr />
          <p style="text-align: justify;">sirFAST is a read mapper that is
            designed to map short reads to reference genome with a special
            emphasis on the discovery of structural variation and segmental
            duplications. sirFAST maps short reads with respect to user defined
            error threshold. This manual,
            describes how to choose the parameters and tune sirFAST with respect
            to the library settings. sirFAST is designed to find <b><u>'all'</u></b> 
            mappings for a given set of reads, however it can return one "best"
            map location if the relevant parameter is invoked.</p>
          <p style="text-align: justify;"><span style="font-weight: bold;">NOTE:</span>
            sirFAST is developed for Complete Genomics (CG), thus requires all reads to be at
            the same length. For paired-end reads, lengths of mates may be
            different from each other, but each "side" should have a uniform
            length<span style="font-weight: bold;"><span style="color: red;"><strong><span
                    style="color: black;">.</span></strong></span></span><span style="text-decoration: underline;"><span
                style="font-weight: bold;"><span style="color: red;"> <br />
                </span></span></span></p>
          <p style="text-align: justify;"><span style="text-decoration: underline;"><span
                style="font-weight: bold;"><span style="color: red;"></span></span></span><span
              style="color: red;"></span><br />
          </p>
          <p style="text-align: justify;"> </p>
          <hr />
          <h2>Table of Contents</h2>
          <ul>
            <li><a href="#sample">Sample Set</a><br />
            </li>
            <li><a href="#general">General</a></li>
            <li><a href="#indexing">Indexing</a>
              <ul>
                <li><a href="#indexing-single-mode">Single Genome Mode</a></li>
              </ul>
            </li>
            <li><a href="#mapping">Mapping</a>
              <ul>
                <li><a href="#mapping-single-mode">Single-end Reads - Single
                    Mode</a></li>
                <li><a href="#mapping-pair-end">Paired-end Reads</a></li>
                <li><a href="#mapping-pair-end-disc">Discordant Paired-end Reads</a></li>
              </ul>
            </li>
            <li><a href="#output">Output Format</a> </li>
          </ul>
          <hr />
          <h2><a name="smaple">Sample Set</a></h2>
		We have provided a simple test to make sure sirFAST is working on your system properly. The sample files 
		are in the Sample_Data folder. We provide a small 
		template reference genome (sample_genome.fa) and a small test read sets (sample_reads_e0). The sample read set is generated
		with no error.
		<p>
			The first step is to make an index of the refernece genome as follow:
				<p><code> $&gt;./sirfast --index Sample_Data/sample_genome.fa<br /> </code> </p>
			
		</p>  
			The output you should see on your screen is:
			<code>
				<p>Generating Index from Sample_Data/sample_genome.fa </p>
 				<p>- chr1 </p>
				<p>DONE in 0.31s!</p>
			</code>
		<p>
                        The second step is to make an index of the refernece genome as follow:
                                <p><code> $&gt;./sirfast --search Sample_Data/sample_genome.fa --seq Sample_Data/sample_reads_e0 -e 0 -o Sample_Data/out.sam <br /> </code> </p>
                        
                </p>  
                        The output you should see on your screen is:
                        <pre>
WARNING: Sample_Data/sample_genome.fa.fai not found, the SAM file(s) will not have a header.
You can generate the .fai file using samtools. Please place it in the same directory with the index to enable SAM headers.
=== 1996 sequences are read in 0.00. (0 discarded) [Mem:323.33 M] ===
-------------------------------------------------------------------------------------------------------------
|       Seq. Name |    Loading Time |    Mapping Time | Memory Usage(M) |  Total Mappings |    Mapped reads |
-------------------------------------------------------------------------------------------------------------
|            chr1 |            0.09 |            0.03 |          328.00 |            3728 |            1996 |
-------------------------------------------------------------------------------------------------------------
=== No more reads for mapping: finishing mapping ===
Total:            0.09              0.03

Post Processing Time:               0.02 
Total Time:                         0.15
Total No. of Reads:                 1996
Total No. of Mappings:              3728
Avg No. of locations verified:         0
			</pre>

			This command generates two files. One is the Sample_Data/out.sam which contains the mapping of the reads and unmapped files 
			which containts the unmapped reads.
 	  <h2><a name="general">General</a></h2>
          Please download the latest version from our download page and then
          unzip the downloaded file. Run 'make' to build sirFAST. <br />
          <br />
              sirFAST <br />
          <ol>
            <li>    generates an index of the reference genome(s) and </li>
            <li>    maps the reads to reference genome. </li>
          </ol>
          <span style="font-weight: bold;">Requirements:</span><br />
          <ul>
            <li><a href="http://zlib.net/">zlib</a> for the ability to read
              compressed FASTQ and write compressed SAM files.</li>
            <li>C compiler (sirFAST is developed with gcc versions &gt; 4.1.2)</li>
          </ul>
          <br />
          <span style="font-weight: bold;"> Building:<br />
            <br />
          </span>
          <p><a> On Unix/Linux systems, we recommend using GNU gcc version &gt;
              4.1.2 as your compiler and type 'make' to build. <br />
              Example:<br />
            </a></p>
          <p> </p>
          <a> </a>
          <pre class="manual"><a>linux&gt; make<br />	<br />gcc -c -O3 baseFAST.c -o baseFAST.o<br />gcc -c -O3 CommandLineParser.c -o CommandLineParser.o<br />gcc -c -O3 Common.c -o Common.o<br />gcc -c -O3 HashTable.c -o HashTable.o<br />gcc -c -O3 SirFAST.c -o SirFAST.o<br />gcc -c -O3 Output.c -o Output.o<br />gcc -c -O3 Reads.c -o Reads.o<br />gcc -c -O3 RefGenome.c -o RefGenome.o<br />gcc baseFAST.o CommandLineParser.o Common.o HashTable.o SirFAST.o Output.o Reads.o RefGenome.o  <br />	-o sirFAST -lz -lm<br />rm -rf *.o<br />	</a></pre>
          <a> </a> <br />
          <b>Parallelization:</b> The best way to optimize sirFAST is to split
          the reads into chunks that fit into the memory of the cluster nodes,
          and implement an MPI wrapper in an 
		<a style="font-style: italic;" href="http://en.wikipedia.org/wiki/Embarrassingly_parallel">embarrassingly parallel</a> fashion. We recommend the following criteria to split
          the reads:<br />
          <br />
          <ol>
            <li><span style="font-style: italic;">Single End Mode:</span> The
              number of reads should be approximately <b>((M-600)/(4*L))
                million</b> where M is the size of the memory for the cluster
              node (in megabytes) and L is the read length. If you have more
              nodes, you can make the chunks smaller to use the nodes
              efficiently. For example, if the library length is 50bp and the
              memory of nodes is 2 GB, each chunk should contain
              (2000-600)/(4*50)= 7 million reads. </li>
            <li><span style="font-style: italic;">Paired End Mode:</span> The
              number of reads in each file should not exceed 1 million (500,000
              pairs), however chunk size of 500,000 reads (250,000 pairs)  is
              recommended.</li>
          </ol>
          <br />
          To see the list of options, use "-h" or "--help".<br />
          To see the version of sirFAST, user "-v" or "--version". <br />
          <br />
          <hr />
          <h2><a name="indexing">Indexing</a></h2>
          sirFAST's indices can be generated in one mode (single). In
          single mode, sirFAST indexes a <em>fasta</em> file (which may contain
          one or more reference genomes). <br />
          <br />
          By default sirFAST uses the window size of 10 characters to generate
          its index. Please be advised that if you do not choose the window size
          carefully, you <b>will</b> lose sensitivity.<br />
          <br />
          <h3>Single Genome Mode:</h3>
          <a name="indexing-single-mode">To index a reference genome like
            "refgen.fasta" run the following command: </a>
          <pre class="manual"><a name="indexing-single-mode">$&gt;./sirfast --index refgen.fasta<br /></a></pre>
          <a name="indexing-single-mode">Upon the completion of the indexing
            phase, you can find "refgen.fasta.index" in the same directory as
            "refgen.fasta". sirFAST uses a window size of 12 (default) to make
            the index of the genome, this windows size can be modified with
            "--ws". There is a restriction on the maximum of the window size as
            the window size directly affects the memory usage. </a>
          <pre class="manual"><a name="indexing-single-mode">$&gt;./sirfast --index refgen.fasta --ws 13<br /></a></pre>
          <hr /> <a name="mapping"> </a>
          <h2><a name="mapping">Mapping</a></h2>
          <a name="mapping">sirFAST can map single-end reads and paired-end reads
            to a reference genome. sirFAST can map in either single or batch
            mode. In single mode, it only maps to one index. In batch mode, it
            maps to a list of indices. sirFAST supports <em>tsv</em> formats.<br />
            <br />
          </a><a name="mapping-single-mode"> </a>
          <h3><a name="mapping-single-mode">Single-end Reads - Single Mode</a></h3>
          <a name="mapping-single-mode">To map single reads to a reference
            genome in single mode, run the following command. Use "--seq" to
            specify the input file. refgen.fa and refgen.fa.index should be in
            the same folder. You can load a multi-sequence FASTA file as the
            reference genome. </a>
          <pre class="manual"><a name="mapping-single-mode">$&gt;./sirfast --search refgen.fa --seq reads.tsv<br /></a></pre>
          <a name="mapping-single-mode">The reported locations will be saved
            into "output" by default. If you want to save it somewhere else, use
            "-o" to specify another file. sirFAST can report the unmapped reads
            in fasta/fastq format. </a>
          <pre class="manual"><a name="mapping-single-mode">$&gt;./sirfast --search refgen.fasta --seq reads.tsv -o my.map <br /></a></pre>
          <a name="mapping-single-mode">By default, sirFAST reports all the
            locations per read. If you need one <span style="font-weight:
              bold;">"best"</span> mapping add the "--best" parameter to the
            command line:<br />
          </a>
          <pre class="manual"><a name="mapping-single-mode">$&gt;./sirfast --search refgen.fasta --seq reads.tsv -e 3 --best<br /></a></pre>
          <a name="mapping-pair-end"> </a>
          <h3><a name="mapping-pair-end">Paired-end Reads</a></h3>
          <a name="mapping-pair-end">To map paired-end reads, use "--pe" option.
            The mapping can be done in single/batch mode. If the reads are in
            two different files, you have to use "--seq1/--seq2" to indicate the
            files. If the reads are interleaved, use "--seq" to indicated the
            file. The distance allowed between the paired-end reads should be
            specified with "--min" and "--max". "--min" and "--max" specify the
            minmum and maximum of the inferred size (the distance between outer
            edges of the mapping mates). </a>
          <pre class="manual"><a name="mapping-pair-end">$&gt;./sirfast --search refgen.fasta --pe --seq reads.tsv --min 150 --max 250 <br /></a></pre>
          <a name="mapping-pair-end-disc"> </a>
          <h3><a name="mapping-pair-end-disc">Discordant Mapping</a></h3>
          <a name="mapping-pair-end-disc">sirFAST can report the discordant
            mapping for use of Variation Hunter (<b> Still on progress! </b>). The --min and --max optiopns
            will define the minimum and maximum inferred size for concordant
            mapping. </a>
          <pre class="manual"><a name="mapping-pair-end-disc">$&gt;./sirfast --search refgen.fasta --pe --discordant-vh --seq reads.tsv --min 50 --max 75<span
style="font-family: arial,sans-serif;"><br /></span></a></pre>
          <a name="output"><br />
          </a>
          <hr style="width: 100%; height: 2px;" /><a name="output"><br />
          </a>
          <h2><a name="output">Parameters</a></h2>
          <a name="output"><br />
            <span style="font-weight: bold;">    General Options:</span><br />
            <br />
                   <br />
          </a>
          <table cellspacing="2" cellpadding="2" border="1" style="text-align:
            left; width: 846px; height: 62px; margin-left: auto; margin-right:            auto; background-color: rgb(211, 211, 211);">
            <tbody>
              <tr>
                <td style="vertical-align: top;">-v|--version </td>
                <td style="vertical-align: top;">Shows the current
                  version.                                                                                                                                       
                  <br /> </td>
              </tr>
              <tr>
                <td style="vertical-align: top;">-h</td>
                <td style="vertical-align: top;">Shows the help screen.<br />
                </td>
              </tr>
            </tbody>
          </table>
          <a name="output">     <br />
            <br />
                <span style="font-weight: bold;">Indexing Options:</span><br />
            <br />
                   <br />
          </a>
          <table cellspacing="2" cellpadding="2" border="1" style="text-align:
            left; width: 846px; height: 62px; margin-left: auto; margin-right:            auto; background-color: rgb(211, 211, 211);">
            <tbody>
              <tr>
                <td style="vertical-align: top;">--index [file]  <br />
                </td>
                <td style="vertical-align: top;">Generate an index from the
                  specified fasta file.</td>
              </tr>
              <tr>
                <td style="vertical-align: top;">-ws [int] </td>
                <td style="vertical-align: top;">Set window size for indexing
                  (default:10 max:14).</td>
              </tr>
            </tbody>
          </table>
          <a name="output">     <br />
            <br />
                <span style="font-weight: bold;">Searching Options:</span><br />
            <br />
          </a>
          <table cellspacing="2" cellpadding="2" border="1" style="text-align:
            left; width: 900px; height: 631px; margin-left: auto; margin-right:            auto; background-color: rgb(211, 211, 211);">
            <tbody>
              <tr>
                <td style="vertical-align: top;">--search [file]   <br />
                </td>
                <td style="vertical-align: top; background-color: rgb(204, 204,
                  204);">Search the specified genome. Index file should be in
                  same directory as the fasta file.</td>
              </tr>
              <tr>
                <td style="vertical-align: top;">--pe<br />
                </td>
                <td style="vertical-align: top;"> Search will be done in
                  paired-end mode</td>
              </tr>
              <tr>
                <td style="vertical-align: top;">--seq [file]</td>
                <td style="vertical-align: top;"> Input sequences in tsv 
                  format [file]. If pairend reads are interleaved, use this
                  option.</td>
              </tr>
              <tr>
                <td style="vertical-align: top;">--seq1 [file]</td>
                <td style="vertical-align: top;"> Input sequences in tsv
                  format [file] (First file). Use this option to indicate the
                  first file of paired-end reads</td>
              </tr>
              <tr>
                <td style="vertical-align: top;">--seq2 [file]</td>
                <td style="vertical-align: top;">Input sequences in tsv
                  format [file] (Second file). Use this option to indicate the
                  second file of paired-end reads.</td>
              </tr>
	      <tr>
		<td style="vertical-align: top;">--t [int]</td>
                <td style="vertical-align: top;">Indicates the number of reads in a chunk which is used to map in each batch.
		This value is used for the ReadSlice version of the code when the number of mapping reads is too large. We split the files
		to chunks and map each one seperate in sequantial manner (default: 100,000 reads)   </td>
	      </tr>
	      <tr>
		<td style="vertical-align: top;">--f [int]</td>
                <td style="vertical-align: top;"> <p>Indicates the read format. <br>
			1 --> first read format 5-10-10-10(first mate)&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;      10-10-10-5 (second mate) <br>
			2 --> second read format 10-9-N-10(first mate)&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;      10-N-9-10  (second mate) <br>
			default: 2  </p>
		</td>	
	      </tr>	
              <tr>
                <td style="vertical-align: top;">-o [file] </td>
                <td style="vertical-align: top;">Output of the mapped sequences
                  (SAM format). The default is "output".</td>
              </tr>
              <tr>
                <td style="vertical-align: top;">-u [file]<br />
                </td>
                <td style="vertical-align: top;">FASTA/FASTQ file for the
                  unmapped sequences. The default is "unmapped".</td>
              </tr>
              <tr>
                <td style="vertical-align: top;">-e [int]</td>
                <td style="vertical-align: top;">Maximum allowed edit distance
                  (default 4% of the read length). Note that although the
                  current version is limited with up to 4+4 indels, it supports
                  any number of substitution errors. </td>
              </tr>
              <tr>
                <td style="vertical-align: top;">--min [int] </td>
                <td style="vertical-align: top;">Min inferred distance allowed
                  between two pairend sequences.</td>
              </tr>
              <tr>
                <td style="vertical-align: top;">--max [int]</td>
                <td style="vertical-align: top;">Max inferred distance allowed
                  between two pairend sequences.</td>
              </tr>
              <tr>
                <td style="vertical-align: top;"> --discordant-vh</td>
                <td style="vertical-align: top;">To return all discordant map
                  locations ready for the Variation Hunter program, and OEA map
                  locations ready for the NovelSeq.<br />
                </td>
              </tr>
              <tr>
                <td style="vertical-align: top;">--best</td>
                <td style="vertical-align: top;">Return "best" location only
                  (single-end mode).</td>
              </tr>
              <tr>
                <td style="vertical-align: top;">--seqcomp</td>
                <td style="vertical-align: top;">Indicates that the input
                  sequences are compressed (gz).</td>
              </tr>
              <tr>
                <td style="vertical-align: top;">--outcomp</td>
                <td style="vertical-align: top;">Indicates that output file
                  should be compressed (gz).</td>
              </tr>
              <tr>
                <td style="vertical-align: top;">--maxoea [int]<br />
                </td>
                <td style="vertical-align: top;">Max number of One End Anchored
                  (OEA) returned for each read pair. Minimum of 100 is
                  recommendded for NovelSeq use.<br />
                </td>
              </tr>
              <tr>
                <td style="vertical-align: top;">--maxdis [int]<br />
                </td>
                <td style="vertical-align: top;">Max number of discordant map
                  locations returned for each read pair.<br />
                </td>
              </tr>
              <tr>
                <td style="vertical-align: top;">--crop [int]</td>
                <td style="vertical-align: top;">Crop the input reads at
                  position [int].</td>
              </tr>
              <tr>
                <td style="vertical-align: top;">--sample [string]<br />
                </td>
                <td style="vertical-align: top;">Sample name to be added to the
                  SAM header (optional).<br />
                </td>
              </tr>
            </tbody>
          </table>
          <br />
           <br />
          <br />
          <hr style="width: 100%; height: 2px;" />
          <h2><br />
          </h2>
          <h2><a name="output">Output Files</a></h2>
          <h2> </h2>
          <div style="text-align: justify;"><span style="font-weight: bold;
              font-style: italic;">Single-End Mode: </span>In the single-end
            mode sirFAST will generate two files as specified by the "-o" and
            "-u" parameters. Default filenename if the "-o" parameter is not
            specified is "output"; and default filename for the "-u" parameter
            is "unmapped". <span style="font-weight: bold;"><br />
            </span></div>
          <span style="font-weight: bold;"> </span><span style="font-weight:
            bold;">
            <ul>
              <li style="font-weight: normal; text-align: justify;"><span style="font-style: italic;">output
                  file ("-o"):  </span><span style="font-style: italic;">Contains
                  the map locations of the sequences in the specified genome in
                  SAM format. sirFAST returns <span style="font-style: italic; font-weight: bold;">all</span>
                  possible map locations within the given edit distance ("-e")
                  by default. If the "--best" parameter is invoked, then it will
                  select one "best" location that has the minimum edit distance
                  to the genome.</span></li>
              <span style="font-weight: normal;"> </span>
              <li style="font-weight: normal;"><span style="font-style: italic;">unmappped
file



                  ("-u"):</span> Contains the unmapped reads in FASTQ or FASTA
                format, depending on the format of the input sequences.</li>
            </ul>
          </span><span style="font-weight: bold;">
            <div style="text-align: justify;"><span style="font-weight: bold;">Paired-End



                and Matepair Modes</span>:<span style="font-style: italic;"><span
                  style="font-style: italic; font-weight: normal;"> </span></span><span
                style="font-weight: normal;">In paired-end and matepair modes,
                sirFAST will generate a SAM file in the paired-end mode that will
                store best mapping locations while utilizing the paired-end span
                information. In addition, it will generate a DIVET file and and
                OEA file (SAM format). See below:</span><span style="font-style:
                italic;"><span style="font-weight: bold;"><span style="font-weight: bold;"><br />
                    <span style="font-weight: bold;"><span style="font-weight:
                        bold;"> </span></span><br />
                    <span style="font-weight: bold;"><span style="font-weight:
                        bold;"> <span style="font-weight: bold; font-style:
                          italic;"><span style="font-weight: bold;">
                            <ul>
                              <li style="font-weight: normal;"><span style="font-style: italic;">output
                                  file ("-o"):</span>  Contains the map
                                locations of the sequences in the specified
                                genome in SAM format. This file will include:<br />
                              </li>
                            </ul>
                          </span></span></span></span></span></span></span></div>
            <span style="font-weight: bold;">
              <ul>
                <ul style="font-weight: normal; text-align: justify;">
                  <li>If a read pair can be mapped concordantly, the "best"
                    (minimum total edit distance and minimum differential from
                    the average span) map location for the pair.</li>
                  <li>If the read pair can not be mapped concordantly, again,
                    the "best" (minimum total edit distance and minimum
                    differential from the average span) map location for the
                    pair.</li>
                </ul>
                <li style="font-weight: normal; text-align: justify;"><span style="font-style: italic;">unmapped
                    file ("-u"):</span> Contains the orphan (both ends unmapped)
                  reads in FASTQ or FASTA format, depending on the format of the
                  input sequences.<br />
                </li>
                <li style="font-weight: normal; text-align: justify;"><span style="font-style: italic;">output.DIVET.vh
                    file ("-o" option changes the prefix "output"): </span>This
                  file includes <span style="font-style: italic; font-weight: bold;">all</span>
                  possible map locations for the read pairs that cannot be
                  concordantly mapped. This file can be loaded by <a href="http://compbio.cs.sfu.ca/strvar.htm">VariationHunter</a>
                  tool for structural variation discovery (<b> Still on progress! </b>).<br />
                </li>
                <li style="text-align: justify;"><span style="font-style:
                    italic; font-weight: normal;">output_OEA file: </span><span
                    style="font-weight: normal;">Contains the OEA
                    (One-End-Anchored) reads (paired-end reads where only one
                    read can be mapped to the genome). The output is in SAM
                    format, contains the map location of read that can be mapped
                    to the genome. The unmapped reads of an OEA read pair are
                    not reported in separate lines; instead the sequence and
                    quality information is given in the line that specifies the
                    map location of the mapped read. We use optional fields NS
                    and NQ to specify the unmapped sequence and unmapped quality
                    information. This file can be loaded by </span><a href="http://compbio.cs.sfu.ca/strvar.htm"
                    target="_blank">NovelSeq</a> <span style="font-weight: normal;">tool
                    for novel sequence discovery (<b> Still on progress! </b>), however format conversion
                    might be required; please see the </span><span style="font-weight: bold;"><a
                      href="http://compbio.cs.sfu.ca/strvar.htm" target="_blank">NovelSeq</a><span
                      style="font-weight: normal;"> documentation.</span></span></li>
                <ul>
                  <li style="text-align: justify;"><span style="font-weight:
                      bold;"><span style="font-weight: normal;"><span style="font-weight: bold;"><span
                            style="font-weight:
                            bold;"><span style="font-weight: normal;"><span style="font-style: italic;">NOTE:
                                sirFAST will report <span style="font-weight: bold;
                                  font-style: italic;">many </span></span></span></span></span></span></span><span
                      style="font-weight: normal;"><span style="font-weight:
                        normal;"><span style="font-style: italic;"><span style="font-style: italic;">(up
                            to 100 by default)</span></span></span></span><span
                      style="font-weight: bold;"><span style="font-weight:
                        normal;"><span style="font-weight: bold;"><span style="font-weight: bold;"><span
                              style="font-weight:
                              normal;"><span style="font-style: italic;"><span style="font-weight: bold; font-style: italic;">
                                </span>possible map locations for the "mapped"
                                read of OEA  matepais. This will generate a
                                large file due to repeats and duplications. This
                                file can be limited through the --maxoea
                                parameter ( <b> Still in progress! <b>).<br />
                              </span></span></span></span></span></span></li>
                </ul>
              </ul>
            </span><span style="font-weight: bold;"><br />
              <br />
              <br />
              <hr style="width: 100%; height: 2px;" /> </span>
            <h2><br />
            </h2>
            <h2><a name="output">Output Format</a></h2>
          </span><span style="font-weight: bold;">
            <div style="text-align: justify;"><span style="font-weight: normal;">sirFAST
		mapping	output format is in SAM format. For detail about the definition
                of the fields please refer to </span><a style="font-weight: normal;"
                href="http://samtools.sourceforge.net/SAM1.pdf">SAM Manual</a><span
                style="font-weight: normal;">. We have not implemented "MQUAL"
                field yet. All locations of discordant paired-end reads will be
                reported in DIVET format as required by the </span><a style="font-weight:
                normal;" href="http://compbio.cs.sfu.ca/strvar.htm">VariationHunter</a><span
                style="font-weight: normal;"> package. Unmapped reads (or,
                "orphan" read pairs in the PE mode) will be outputted in FASTQ
                or FASTA format, depending on the input sequence file format.</span><br />
            </div>
            <span style="font-weight: bold;"><br />
            </span></span></div>
        <br />
      </div>
    </div>
  </body>
</html>
